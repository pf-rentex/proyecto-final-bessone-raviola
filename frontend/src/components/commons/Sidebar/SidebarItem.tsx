import React from 'react';
import { Accordion } from '../Accordion/Accordion';

const SidebarItem = ({ item }: any) => {
    return (
        <li className="px-3 py-5">
            {item?.children ? (
                <Accordion title={item.title}>
                    <ul className="font-thin space-y-5 pl-3 pr-1">
                        {item.children.map((child: any, index: number) => (
                            <li
                                key={index}
                                className="flex justify-between cursor-pointer select-none"
                                onClick={() => child?.callback()}
                            >
                                {child.title}
                                {child?.icon && child.icon}
                            </li>
                        ))}
                    </ul>
                </Accordion>
            ) : (
                <div
                    className="flex justify-between items-center cursor-pointer select-none"
                    onClick={() => item?.callback()}
                >
                    {item.title}
                    {item?.icon && item.icon}
                </div>
            )}
        </li>
    );
};

export default SidebarItem;
