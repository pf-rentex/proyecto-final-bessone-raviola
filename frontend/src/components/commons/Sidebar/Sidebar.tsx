import React, { useRef } from 'react';
import { BiSearch } from 'react-icons/bi';
import { BsPlusCircleFill } from 'react-icons/bs';
import { Badge } from '../Badge/Badge';
import { useClickedOutside } from '../../../utils';
import SidebarItem from './SidebarItem';
import { AiOutlinePlus } from 'react-icons/ai';
import { useNavigate } from 'react-router-dom';

interface ISidebarProps {
    isOpen: boolean;
    setIsOpen: Function;
}

export interface ISidebarItem {
    title: string;
    icon?: React.ReactNode;
    callback?: () => void;
    children?: { title: string; callback?: () => void; icon: React.ReactNode }[];
}

const Sidebar = ({ isOpen, setIsOpen }: ISidebarProps) => {
    const wrapperRef = useRef(null);
    useClickedOutside(wrapperRef, () => isOpen && setIsOpen(false));
    const navigate = useNavigate();

    const items: ISidebarItem[] = [
        {
            title: 'Solicitudes',
            callback: () => {
                navigate('/rent/requests');
                setIsOpen(false);
            },
            icon: <Badge>4</Badge>,
        },
        {
            title: 'Publicaciones',
            callback: () => {
                navigate('/properties/search');
                setIsOpen(false);
            },
            icon: <BsPlusCircleFill />,
        },
        {
            title: 'Reclamos',
            callback: () => {
                navigate('/claims');
                setIsOpen(false);
            },
            icon: <Badge>4</Badge>,
        },
        {
            title: 'Contratos',
            children: [
                {
                    title: 'Crear',
                    callback: () => {
                        navigate('/contracts/create');
                        setIsOpen(false);
                    },
                    icon: <AiOutlinePlus className="w-5 h-5" />,
                },
                {
                    title: 'Ver todos',
                    callback: () => {
                        navigate('/contracts');
                        setIsOpen(false);
                    },
                    icon: <BiSearch className=" w-5 h-5" />,
                },
            ],
        },
        {
            title: 'Propiedades',
            callback: () => {
                navigate('/properties');
                setIsOpen(false);
            },
            icon: <BsPlusCircleFill />,
        },
    ];

    return (
        <div
            data-testid={'sidebar'}
            ref={wrapperRef}
            className={`flex flex-col w-8/12 lg:w-5/12 xl:w-3/12 h-full fixed z-30 bg-gradient-to-b from-bg-gradient-11 to-bg-gradient-12 text-white transition duration-300 transform ${
                isOpen ? 'translate-x-0 ' : '-translate-x-full'
            }`}
        >
            <div className={`bg-alt w-full p-3  ${!isOpen ? 'translate-x-0 ' : '-translate-x-full'}`}>
                <h1 className="text-center text-3xl font-semibold">Dashboard</h1>
            </div>
            <div className={`p-4 ${!isOpen ? 'translate-x-0 ' : '-translate-x-full'}`}>
                <ul className="divide-y text-xl">
                    {items.map((item: ISidebarItem, index) => (
                        <SidebarItem key={index} item={item}></SidebarItem>
                    ))}
                </ul>
            </div>
        </div>
    );
};

export default Sidebar;
