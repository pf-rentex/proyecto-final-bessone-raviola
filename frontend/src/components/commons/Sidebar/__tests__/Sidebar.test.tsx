import React from "react";
import {fireEvent, render} from "@testing-library/react";
import Sidebar from "../Sidebar";

describe("Sidebar", () => {

  test("renders component", () => {
    const wrapper = render(<Sidebar isOpen={false} setIsOpen={() => {}}/>);

    expect(wrapper.container).toBeInTheDocument();
    expect(wrapper.container).toMatchSnapshot();
  });

  test("Sidebar has corresponding classes when closed and open", () => {
    const wrapper = render(<Sidebar isOpen={false} setIsOpen={() => {}}/>);

    expect(wrapper.getByTestId('sidebar')).toHaveClass('-translate-x-full');

    wrapper.rerender(<Sidebar isOpen={true} setIsOpen={() => {}}/>);

    expect(wrapper.getByTestId('sidebar')).toHaveClass('translate-x-0');
  });

  test('Sidebar initially open is closed by clicking outside', () => {
    const setIsOpen = jest.fn();
    const wrapper = render(<Sidebar isOpen={true} setIsOpen={setIsOpen}/>);

    expect(wrapper.getByTestId('sidebar')).toHaveClass('translate-x-0');
    fireEvent.mouseDown(document);

    expect(setIsOpen).toHaveBeenCalledWith(false);
  });
});
