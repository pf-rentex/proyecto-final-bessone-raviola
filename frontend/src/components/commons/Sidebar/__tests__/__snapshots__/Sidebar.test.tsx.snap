// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders component 1`] = `
<div>
  <div
    class="flex flex-col w-8/12 lg:w-5/12 xl:w-3/12 h-full fixed z-30 bg-gradient-to-b from-bg-gradient-11 to-bg-gradient-12 text-white transition duration-300 transform -translate-x-full"
  >
    <div
      class="bg-alt w-full p-3  translate-x-0 "
    >
      <h1
        class="text-center text-3xl font-semibold"
      >
        Dashboard
      </h1>
    </div>
    <div
      class="p-4 translate-x-0 "
    >
      <ul
        class="divide-y text-xl"
      >
        <li
          class="px-3 py-5"
        >
          <div
            class="flex justify-between items-center cursor-pointer select-none"
          >
            Solicitudes
            <span
              class="inline-flex items-center justify-center px-2 py-1 text-xs font-bold leading-none bg-red-500 rounded-full"
            >
              4
            </span>
          </div>
        </li>
        <li
          class="px-3 py-5"
        >
          <div
            class="flex flex-col"
          >
            <button
              class="box-border appearance-none cursor-pointer focus:outline-none flex items-center justify-between"
              data-testid="button"
            >
              <p
                class="inline-block text-footnote light"
              >
                Mis inmuebles
              </p>
              <svg
                class="transform duration-300 ease inline-block text-alt w-7 h-7"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zM273 369.9l135.5-135.5c9.4-9.4 9.4-24.6 0-33.9l-17-17c-9.4-9.4-24.6-9.4-33.9 0L256 285.1 154.4 183.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L239 369.9c9.4 9.4 24.6 9.4 34 0z"
                />
              </svg>
            </button>
            <div
              class="overflow-hidden transition-max-height duration-300 ease-in-out"
              style="max-height: 0px;"
            >
              <div
                class="pt-5"
              >
                <ul
                  class="font-thin space-y-5"
                >
                  <li
                    class="flex justify-between"
                  >
                    Opción 1
                  </li>
                  <li
                    class="flex justify-between"
                  >
                    Opción 2
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </li>
        <li
          class="px-3 py-5"
        >
          <div
            class="flex flex-col"
          >
            <button
              class="box-border appearance-none cursor-pointer focus:outline-none flex items-center justify-between"
              data-testid="button"
            >
              <p
                class="inline-block text-footnote light"
              >
                Contratos
              </p>
              <svg
                class="transform duration-300 ease inline-block text-alt w-7 h-7"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zM273 369.9l135.5-135.5c9.4-9.4 9.4-24.6 0-33.9l-17-17c-9.4-9.4-24.6-9.4-33.9 0L256 285.1 154.4 183.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L239 369.9c9.4 9.4 24.6 9.4 34 0z"
                />
              </svg>
            </button>
            <div
              class="overflow-hidden transition-max-height duration-300 ease-in-out"
              style="max-height: 0px;"
            >
              <div
                class="pt-5"
              >
                <ul
                  class="font-thin space-y-5"
                >
                  <li
                    class="flex justify-between"
                  >
                    Ver contratos
                    <svg
                      class="text-alt w-7 h-7"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M10,18c1.846,0,3.543-0.635,4.897-1.688l4.396,4.396l1.414-1.414l-4.396-4.396C17.365,13.543,18,11.846,18,10 c0-4.411-3.589-8-8-8s-8,3.589-8,8S5.589,18,10,18z M10,4c3.309,0,6,2.691,6,6s-2.691,6-6,6s-6-2.691-6-6S6.691,4,10,4z"
                      />
                    </svg>
                  </li>
                   
                  <li
                    class="flex justify-between"
                  >
                    Crear nuevo 
                    <svg
                      class="text-alt w-7 h-7"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 16 16"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8.5 4a.5.5 0 00-1 0v3.5H4a.5.5 0 000 1h3.5V12a.5.5 0 001 0V8.5H12a.5.5 0 000-1H8.5V4z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </li>
        <li
          class="px-3 py-5"
        >
          Reclamos
        </li>
      </ul>
    </div>
  </div>
</div>
`;
